src/mbgl/renderer/painter_fill.cpp

```diff
@@ -90,6 +90,26 @@ void Painter::renderFill(FillBucket& bucket, const FillLayer& layer, const TileI
             patternShader->u_patternmatrix_a = patternMatrixA;
             patternShader->u_patternmatrix_b = patternMatrixB;
 
+            std::array<int, 2> imageSizeScaledA = {{
+                (int)((*posA).size[0] * properties.pattern.value.fromScale),
+                (int)((*posA).size[1] * properties.pattern.value.fromScale)
+            }};
+            std::array<int, 2> imageSizeScaledB = {{
+                (int)((*posB).size[0] * properties.pattern.value.toScale),
+                (int)((*posB).size[1] * properties.pattern.value.toScale)
+            }};
+
+            int tileSize = 512;
+
+            float offsetAx = ((tileSize % imageSizeScaledA[0]) * id.x) / (float)imageSizeScaledA[0];
```

Comment:
It looks like `x` is wrapped in -js: https://github.com/mapbox/mapbox-gl-js/blob/25b02a05d29130702e8f66e6a2bfe91333ac010f/js/source/tile_coord.js#L157

but not in -native: https://github.com/mapbox/mapbox-gl-native/blob/11062a01b40cb64c57d1b7b2ef8c0fc39c11d159/src/mbgl/util/tile_cover.cpp#L76-L77

so `x` already has the 'w' part built into it. I opened https://github.com/mapbox/mapbox-gl-native/issues/3529 about that